src/source/sse.rs

```diff
@@ -0,0 +1,196 @@
+    ) -> std::result::Result<Response, http_types::Error> {
+        let mut req = req;
+        // set the headers in the request.
+        for (name, value) in self.header.iter() {
+            req.append_header(name.as_str(), value.as_str());
+        }
+        let resp = next.run(req, client).await?;
+        Ok(resp)
+    }
+}
+
+pub struct Sse {
+    pub config: Config,
+    onramp_id: TremorUrl,
+}
+
+impl onramp::Impl for Sse {
+    fn from_config(id: &TremorUrl, config: &Option<YamlValue>) -> Result<Box<dyn Onramp>> {
+        if let Some(config) = config {
+            let config: Config = Config::new(config)?;
+            Ok(Box::new(Self {
+                config,
+                onramp_id: id.clone(),
+            }))
+        } else {
+            Err("Missing config for SSE onramp".into())
+        }
+    }
+}
+
+#[async_trait::async_trait()]
+impl Onramp for Sse {
+    async fn start(&mut self, config: OnrampConfig<'_>) -> Result<onramp::Addr> {
+        let source = Int::from_config(config.onramp_uid, self.onramp_id.clone(), &self.config);
+        SourceManager::start(source, config).await
+    }
+
+    fn default_codec(&self) -> &str {
+        // What is the default use case? maybe json?
+        "string"
+    }
+}
+
+pub struct Int {
+    uid: u64,
+    config: Config,
+    onramp_id: TremorUrl,
+    event_source: Option<Receiver<surf_sse::Event>>,
+}
+impl std::fmt::Debug for Int {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "SSE")
+    }
+}
+
+impl Int {
+    fn from_config(uid: u64, onramp_id: TremorUrl, config: &Config) -> Self {
+        Int {
+            uid,
+            config: config.clone(),
+            onramp_id,
+            event_source: None,
+        }
+    }
+}
+
+async fn handle_init(
+    url: surf_sse::Url,
+    headers: HashMap<String, String>,
+    tx: Sender<surf_sse::Event>,
+) -> Result<()> {
+    // Create the specific client.
+    let client = surf::client().with(MiddlewareHeader { header: headers });
+    // Ok to block here. This is an indirection from the main thread.
+    let mut event_source = surf_sse::EventSource::with_client(client, url);
+
+    while let Some(event) = event_source.next().await {
+        match event {
+            Ok(event) => {
+                tx.send(event).await?;
+            }
+            Err(e) => {
+                error!("SSE Source Error: {}", e);
+            }
+        };
+    }
+    tx.close();
+    Ok(())
+}
+
+#[async_trait::async_trait()]
+impl Source for Int {
+    fn id(&self) -> &TremorUrl {
+        &self.onramp_id
+    }
+
+    async fn init(&mut self) -> Result<SourceState> {
+        // Neccessary check as Isahc(Curl) client panics if the url isn't complete. Seems like an edge case.
+        // Maybe use the H1-client as the backend. Though Heinz might have a better solution which is impl in rest sink.
+        // How shall I output this error.
+        let url = self.config.url.as_str().parse();
```

Comment:
Shouldn;t this check already be handled by Line 156?

Reply:
It should but there seems to an edge case from my random inputs. In the crate `surf` `Surf::Url` is passed to the underlying http client. But in the implementation of a traits which abstract over the http clients (https://docs.rs/http-client/6.4.1/src/http_client/isahc.rs.html#60 ) an unwrap() panics. So here I verify it beforehand to avoid a thread panic.****

Reply:
Ohhh, now it makes sense, sorry I didn't gasp the comment fully :+1: makes sense!